nested_forward_ad.py

```diff
@@ -0,0 +1,243 @@
+import torch
+from torch.autograd import forward_ad as fwAD
+from torch import Tensor
+from torch.utils._pytree import tree_map
+from torch.testing._internal.common_utils import (
+    TestCase, run_tests, disable_gc, parametrize,
+    instantiate_parametrized_tests
+)
+
+import functools
+import contextlib
+
+from utils import no_dispatch
+from base_tensor import BaseTensor
+
+# WARNING:
+# This class requires https://github.com/pytorch/pytorch/pull/73925 (that was reverted)
+# to properly work with forward AD implementation
+# If you get an error about "Expected this function to only be reached in inference mode"
+# then you don't have that patch!
+
+# This class wraps a pytorch dual Tensor and associates a level to it.
+# This allows to do multi-level forward AD even though pytorch only
+# support one level.
+class ForwardADTensor(BaseTensor):
+    @staticmethod
+    def __new__(cls, dual_t, *, level, ignore_no_grad=False):
+        # Use this to check if the plain object has a forward grad or not while ignoring
+        # all of the torch_dispatch handling
+        with no_dispatch():
+            primal, tangent = fwAD.unpack_dual(dual_t)
+        # Ensure we actually have a dual Tensor
+        assert ignore_no_grad or tangent is not None, "ForwardADTensor can only wrap Tensors with forward gradients"
+        # Ensure that nesting is happening in the right order
+        if isinstance(dual_t, cls):
+            assert dual_t.level < level, "Level ordering is wrong!"
+        res = super().__new__(cls, primal)
+        return res
+
+    def __repr__(self):
+        # Use no_dispatch here to get a plain representation of this Tensor without any of the
+        # torch_dispatch handling
+        with no_dispatch():
+            self_repr = super().__repr__()
+        return f"ForwardADTensor{self.level}({self_repr}, {self.elem!r})"
+
+    def __init__(self, dual_t, *, level, ignore_no_grad=False):
+        self.elem = dual_t
+        self.level = level
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        # Detach is a special case here.
+        # This special case is for the code from autograd that uses shallow_copy_and_detach
+        # (which is rerouted to detach in torch_dispatch) and user code that calls detach
+        # In this case, we want to only detach *one* level of forward grad. Since forward grad
+        # is already handled before getting here, we just want to convert detach into alias before
```

Comment:
ugh, so shallow_copy_and_detach actually has different semantics from detach. So we can't delete shallow_copy_and_detach and replace it with regular detach

Reply:
Not really?
I think all detach in this context have this semantic?

Reply:
For reverse-mode AD: shallow_copy_and_detach() only detaches the top-most level of autograd. The semantics of detach() right now are that it detaches ALL levels of autograd. Does that sound right?

Reply:
To clarify, by "So we can't delete shallow_copy_and_detach and replace it with regular detach", I meant that in pytorch/pytorch we were having a discussion if we can delete shallow_copy_and_detach and replace it with detach() generally and this is a potential reason we cannot

Reply:
Well, one goes through the dispatcher and not the other so yes this kind of crazy behavior will happen :(